File: src/04_coaching/schema.py
"""Schéma de sortie typé de la review de coaching (Pydantic v2).

Expose Review (1-3 forces / 3 erreurs / 2 habitudes) et le JSON-schema dérivé pour le
paramètre `format` d'Ollama (structured output). Chaque Insight porte sa
preuve chiffrée — pas de conseil sans stat.
"""
from __future__ import annotations

import re
from typing import Annotated, Literal, get_args

from pydantic import BaseModel, Field, field_validator, model_validator


class Insight(BaseModel):
    point: str       # affirmation FR ("tu roams trop peu en mid")
    evidence: str    # preuve chiffrée du payload ("roam mid 50% vs 70% challenger")


class Review(BaseModel):
    # 1 à 3 forces : forcer exactement 3 poussait le LLM à remplir avec du filler
    # quand le profil n'a que 1-2 forces réellement saillantes (feedback "trop-vague").
    strengths: Annotated[list[Insight], Field(min_length=1, max_length=3)]
    mistakes: Annotated[list[Insight], Field(min_length=3, max_length=3)]
    habits: Annotated[list[str], Field(min_length=2, max_length=2)]
    next_focus: str
    confidence: Annotated[float, Field(ge=0.0, le=1.0)]


def review_json_schema() -> dict:
    """JSON-schema passé à Ollama `format`. minItems/maxItems contraignent la génération."""
    return Review.model_json_schema()


# --- Review par-game : chaque erreur ancrée sur un moment précis --------------

_CLOCK_RE = re.compile(r"\d+:\d\d")


class AnchoredInsight(Insight):
    """Insight dont la preuve cite un horodatage mm:ss du journal (contrainte
    de schéma : une erreur par-game sans moment cité est invalide par construction)."""

    @field_validator("evidence")
    @classmethod
    def _evidence_has_clock(cls, v: str) -> str:
        if not _CLOCK_RE.search(v):
            raise ValueError("evidence sans horodatage mm:ss")
        return v


class GameInsight(AnchoredInsight):
    """Insight par-game : `point` = la leçon actionnable, `cause` = le POURQUOI
    (mécanisme), `evidence` = la preuve chiffrée + l'horodatage mm:ss (hérité).

    Réponse au feedback (boucle d'éval, 2026-07-08) : l'horodatage seul laissait le
    joueur sans cause de mort (« je sais pas pourquoi je suis mort ») et les forces
    sans explication (« aucune idée de pourquoi »). `cause` force le LLM à expliciter
    le mécanisme (solo 1v1 vs gank, comportement à l'origine d'une force), et ancre les
    forces sur un moment au même titre que les erreurs."""

    cause: str = Field(description=(
        "Le POURQUOI de l'insight : mécanisme de mort (solo 1v1 sans flash, gank 3v1, "
        "overextension) ou comportement à l'origine d'une force. Jamais l'issue."))

    @field_validator("cause")
    @classmethod
    def _cause_nonempty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("cause vide")
        return v


class GameReview(BaseModel):
    """Review d'UNE game. Pas de section habits : une habitude est un pattern
    multi-games, indétectable sur une game isolée (source de vague assurée).
    Forces ET erreurs sont des `GameInsight` : ancrées sur un moment mm:ss + cause
    explicite (cf. GameInsight) — une force sans preuve temporelle ni cause est du
    remplissage vague, on l'exclut plutôt que de la produire."""
    strengths: Annotated[list[GameInsight], Field(max_length=2)]
    mistakes: Annotated[list[GameInsight], Field(min_length=1, max_length=3)]
    next_focus: str
    confidence: Annotated[float, Field(ge=0.0, le=1.0)]


def game_review_json_schema() -> dict:
    return GameReview.model_json_schema()


# --- Review par agents spécialisés ------------------------------------------

AxisKind = Literal["death_positioning", "economy_build"]


class AxisReview(GameReview):
    axis: AxisKind
    label: str


class ChiefSelection(BaseModel):
    """Le chef ne rédige rien : il sélectionne les IDs des sous-agents."""
    summary_insight_id: str
    priority_mistake_ids: Annotated[list[str], Field(min_length=1, max_length=3)]
    strength_insight_ids: Annotated[list[str], Field(max_length=2)]
    next_focus_insight_id: str
    confidence: Annotated[float, Field(ge=0.0, le=1.0)]


class SpecializedGameReview(GameReview):
    """Vue compatible GameReview + analyses complètes dépliables par axe."""
    summary: str
    axes: Annotated[list[AxisReview], Field(min_length=2, max_length=2)]


def chief_selection_json_schema(mistake_ids: list[str], strength_ids: list[str]) -> dict:
    schema = ChiefSelection.model_json_schema()
    props = schema["properties"]
    props["summary_insight_id"]["enum"] = mistake_ids
    props["priority_mistake_ids"]["items"]["enum"] = mistake_ids
    props["next_focus_insight_id"]["enum"] = mistake_ids
    props["strength_insight_ids"]["items"]["enum"] = strength_ids
    return schema


# --- Boucle d'évaluation : annotation des reviews persistées -----------------

TagKind = Literal["asymetrie", "stat-inventee", "profondeur-en-faute",
                 "trop-vague", "non-actionnable", "autre"]
NEG_TAGS: tuple[str, ...] = get_args(TagKind)   # affiché en menu (restera synchronisé)


class FeedbackItem(BaseModel):
    kind: Literal["strength", "mistake", "habit", "focus"]
    index: int                       # position dans sa section (focus = 0)
    useful: bool
    tag: TagKind | None = None        # obligatoire si useful=False (cf. validator)
    note: str | None = None

    @model_validator(mode="after")
    def _tag_required_when_not_useful(self):
        if not self.useful and self.tag is None:
            raise ValueError("tag requis quand useful=False")
        return self


class Feedback(BaseModel):
    ts: str                           # clé = ts de la review annotée
    player: str
    rated_at: str                     # ISO timestamp de l'annotation
    model: str                        # copié de la review (récap par modèle)
    overall_useful: bool | None = None   # non collecté par le flow interactif
    items: list[FeedbackItem]         # items annotés (≤9 ; skips omis)

File: web/cf/src/schema.ts
/** Validation Review — miroir strict de src/04_coaching/schema.py. */

export interface Insight {
  point: string;
  evidence: string;
}

export interface Review {
  strengths: Insight[];
  mistakes: Insight[];
  habits: string[];
  next_focus: string;
  confidence: number;
}

export interface GameInsight extends Insight {
  cause: string;
}

export interface GameReview {
  strengths: GameInsight[];
  mistakes: GameInsight[];
  next_focus: string;
  confidence: number;
}

export const NEG_TAGS = [
  "asymetrie",
  "stat-inventee",
  "profondeur-en-faute",
  "trop-vague",
  "non-actionnable",
  "autre",
] as const;
export type TagKind = (typeof NEG_TAGS)[number];

function isInsight(value: unknown): value is Insight {
  return typeof value === "object" && value !== null
    && typeof (value as Insight).point === "string"
    && typeof (value as Insight).evidence === "string";
}

function isStringArray(value: unknown, length: number): value is string[] {
  return Array.isArray(value)
    && value.length === length
    && value.every((item) => typeof item === "string");
}

function isConfidence(value: unknown): value is number {
  return typeof value === "number"
    && Number.isFinite(value)
    && value >= 0
    && value <= 1;
}

function isGameInsight(value: unknown): value is GameInsight {
  return isInsight(value)
    && typeof (value as GameInsight).cause === "string";
}

export function validateReview(raw: unknown): Review | null {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) return null;
  const value = raw as Record<string, unknown>;
  if (!Array.isArray(value.strengths)
      || value.strengths.length < 1
      || value.strengths.length > 3) return null;
  if (!Array.isArray(value.mistakes) || value.mistakes.length !== 3) return null;
  if (!value.strengths.every(isInsight) || !value.mistakes.every(isInsight)) return null;
  if (!isStringArray(value.habits, 2)) return null;
  if (typeof value.next_focus !== "string") return null;
  if (!isConfidence(value.confidence)) return null;
  return {
    strengths: value.strengths,
    mistakes: value.mistakes,
    habits: value.habits,
    next_focus: value.next_focus,
    confidence: value.confidence,
  };
}

/** Miroir de GameReview : aucune habitude n'est inférée sur une seule partie. */
export function validateGameReview(raw: unknown): GameReview | null {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) return null;
  const value = raw as Record<string, unknown>;
  if (!Array.isArray(value.strengths) || value.strengths.length > 2) return null;
  if (!Array.isArray(value.mistakes)
      || value.mistakes.length < 1
      || value.mistakes.length > 3) return null;
  if (!value.strengths.every(isGameInsight) || !value.mistakes.every(isGameInsight)) return null;
  if (typeof value.next_focus !== "string" || value.next_focus.trim() === "") return null;
  if (!isConfidence(value.confidence)) return null;
  return {
    strengths: value.strengths,
    mistakes: value.mistakes,
    next_focus: value.next_focus,
    confidence: value.confidence,
  };
}

export function reviewJsonSchema(): Record<string, unknown> {
  const insight = {
    type: "object",
    properties: {
      point: { type: "string" },
      evidence: { type: "string" },
    },
    required: ["point", "evidence"],
    additionalProperties: false,
  };
  return {
    type: "object",
    properties: {
      strengths: { type: "array", minItems: 1, maxItems: 3, items: insight },
      mistakes: { type: "array", minItems: 3, maxItems: 3, items: insight },
      habits: {
        type: "array",
        minItems: 2,
        maxItems: 2,
        items: { type: "string" },
      },
      next_focus: { type: "string" },
      confidence: { type: "number", minimum: 0, maximum: 1 },
    },
    required: ["strengths", "mistakes", "habits", "next_focus", "confidence"],
    additionalProperties: false,
  };
}
